raft/src/main/java/org/apache/kafka/raft/QuorumState.java

```diff
@@ -470,15 +470,53 @@ public void transitionToUnattachedVotedState(
      */
     public void transitionToFollower(int epoch, int leaderId, Endpoints endpoints) {
         int currentEpoch = state.epoch();
-        if (localId.isPresent() && leaderId == localId.getAsInt()) {
-            throw new IllegalStateException("Cannot transition to Follower with leader " + leaderId +
-                " and epoch " + epoch + " since it matches the local broker.id " + localId);
+        if (endpoints.isEmpty()) {
+            throw new IllegalArgumentException(
+                String.format(
+                    "Cannot transition to Follower with leader %s and epoch %s without a leader endpoint",
+                    leaderId,
+                    epoch
+                )
+            );
+        } else if (localId.isPresent() && leaderId == localId.getAsInt()) {
+            throw new IllegalStateException(
+                String.format(
+                    "Cannot transition to Follower with leader %s and epoch %s since it matches the local node.id %s",
+                    leaderId,
+                    epoch,
+                    localId
+                )
+            );
         } else if (epoch < currentEpoch) {
-            throw new IllegalStateException("Cannot transition to Follower with leader " + leaderId +
-                " and epoch " + epoch + " since the current epoch " + currentEpoch + " is larger");
-        } else if (epoch == currentEpoch && (isFollower() || isLeader())) {
-            throw new IllegalStateException("Cannot transition to Follower with leader " + leaderId +
-                " and epoch " + epoch + " from state " + state);
+            throw new IllegalStateException(
+                String.format(
+                    "Cannot transition to Follower with leader %s and epoch %s since the current epoch %s is larger",
+                    leaderId,
+                    epoch,
+                    currentEpoch
+                )
+            );
+        } else if (epoch == currentEpoch) {
+            if (isFollower() && state.leaderEndpoints().size() >= endpoints.size()) {
```

Comment:
we're not concerned about the case where we have the same number of endpoints but the endpoints have changed?

Reply:
Yes. Let me file an issue for that. I think that solution is a bit more complicated when I was thinking about this problem. I am convinced that what we really want to do is track Endpoints that came from BEGIN_QUORUM_EPOCH request. In other words, directly from the leader versus endpoints that came indirectly from the RPC responses.

Leader endpoints that come directly from the leader take precedence over leader endpoints that came indirectly from RPC responses.

I have created https://issues.apache.org/jira/browse/KAFKA-17700

Reply:
I think the code is correct because with the current protocol `leaderEndpoints.size() == 1` means that it came from RPC responses (indirectly from other replicas). leaderEndpoints.size() > 0 means it came directly from the leader (BEGIN_QUORUM_EPOCH). Those are the only two possibly.

The last observation or invariant is that a replica (the leader) cannot change its endpoint with a restart. If the leader restarts the leader epoch is guarantee to increase. In other words for the same epoch BEGIN_QUORUM_EPOCH will always send the same leader endpoints.

Again these are all very subtle observation and invariant. I can and will update the code to make this correct but less subtle.

Reply:
Yes, making this less subtle would be useful. Since the collection is small, would it be possible to take a hash code of it or something? (In a follow up JIRA)